node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/InfraDeployerImpl.java

```diff
@@ -0,0 +1,161 @@
+import com.yahoo.component.Version;
+import com.yahoo.config.provision.ApplicationId;
+import com.yahoo.config.provision.Deployment;
+import com.yahoo.config.provision.HostFilter;
+import com.yahoo.config.provision.HostName;
+import com.yahoo.config.provision.HostSpec;
+import com.yahoo.config.provision.InfraDeployer;
+import com.yahoo.config.provision.NodeType;
+import com.yahoo.config.provision.Provisioner;
+import com.yahoo.log.LogLevel;
+import com.yahoo.transaction.Mutex;
+import com.yahoo.transaction.NestedTransaction;
+import com.yahoo.vespa.hosted.provision.Node;
+import com.yahoo.vespa.hosted.provision.NodeRepository;
+import com.yahoo.vespa.hosted.provision.maintenance.InfrastructureVersions;
+import com.yahoo.vespa.service.monitor.DuperModelInfraApi;
+import com.yahoo.vespa.service.monitor.InfraApplicationApi;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+/**
+ * @author freva
+ */
+public class InfraDeployerImpl implements InfraDeployer {
+
+    private static final Logger logger = Logger.getLogger(InfraDeployerImpl.class.getName());
+
+    private final NodeRepository nodeRepository;
+    private final Provisioner provisioner;
+    private final InfrastructureVersions infrastructureVersions;
+    private final DuperModelInfraApi duperModel;
+    private final Map<ApplicationId, InfraApplicationApi> infraApplicationApiByApplicationId;
+
+    @Inject
+    public InfraDeployerImpl(NodeRepository nodeRepository, Provisioner provisioner, DuperModelInfraApi duperModel) {
+        this.nodeRepository = nodeRepository;
+        this.provisioner = provisioner;
+        this.infrastructureVersions = nodeRepository.infrastructureVersions();
+        this.duperModel = duperModel;
+        this.infraApplicationApiByApplicationId = duperModel.getSupportedInfraApplications().stream()
+                .collect(Collectors.toUnmodifiableMap(
+                        InfraApplicationApi::getApplicationId,
+                        Function.identity()));
+    }
+
+
+    @Override
+    public Optional<Deployment> getDeployment(ApplicationId application) {
+        return Optional.ofNullable(infraApplicationApiByApplicationId.get(application)).map(InfraDeployment::new);
+    }
+
+    @Override
+    public List<ApplicationId> getSupportedInfraApplications() {
+        return List.copyOf(infraApplicationApiByApplicationId.keySet());
+    }
+
+    private class InfraDeployment implements Deployment {
+        private final InfraApplicationApi application;
+
+        private boolean prepared = false;
+        private List<Node> candidateNodes;
+        private List<HostSpec> hostSpecs;
+
+        private InfraDeployment(InfraApplicationApi application) {
+            this.application = application;
+        }
+
+        @Override
+        public void prepare() {
+            if (prepared) return;
```

Comment:
I don't like keeping this internal internal state, ideally we would just create `InfraDeployment` once for every app and then just return that, but this is kinda required by the interface:
https://github.com/vespa-engine/vespa/blob/c3667718a63a8703bf62833dcb92b7ad5422d0cc/config-provisioning/src/main/java/com/yahoo/config/provision/Deployment.java#L18